File: solutions/3211. Generate Binary Strings Without Adjacent Zeros/3211.cpp
// Time complexity: O(n^2)
// Space complexity: O(n)
class Solution {
 public:
  vector<string> validStrings(int n) {
    vector<string> ans;
    dfs(n, "", ans);
    return ans;
  }

 private:
  void dfs(int n, string&& s, vector<string>& ans) {
    if (n == 0) {
      ans.push_back(s);
      return;
    }
    if (s.empty() || s.back() == '1') {
      s.push_back('0');
      dfs(n - 1, move(s), ans);
      s.pop_back();
    }
    s.push_back('1');
    dfs(n - 1, move(s), ans);
    s.pop_back();
  }
};

File: solutions/3211. Generate Binary Strings Without Adjacent Zeros/3211.rs
// Time complexity: O(n^2)
// Space complexity: O(n)
impl Solution {
    pub fn valid_strings(n: i32) -> Vec<String> {
        let mut ans = vec![];

        fn dfs(n: i32, s: &mut Vec<char>, ans: &mut Vec<String>) {
            if n == 0 {
                ans.push(s.iter().collect());
                return;
            }
            if s.is_empty() || *s.last().unwrap() == '1' {
                s.push('0');
                dfs(n - 1, s, ans);
                s.pop();
            }
            s.push('1');
            dfs(n - 1, s, ans);
            s.pop();
        }

        dfs(n, &mut vec![], &mut ans);
        ans
    }
}
